File: 2024/Day 08/Python/main.py
"""Advent of Code - 2024 - Day 8"""
from collections import defaultdict
from itertools import permutations

antennas: dict[str, list[tuple[int, int]]] = defaultdict(list)
with open(0, encoding="utf-8") as f:
    city_map = f.read().splitlines()
height = len(city_map)
width = len(city_map[0])
for y, line in enumerate(city_map):
    for x, char in enumerate(line):
        if char != ".":
            antennas[char].append((x, y))

antinodes1 = set()
for locations in antennas.values():
    for (a, b) in permutations(locations, 2):
        pos = (2*a[0]-b[0], 2*a[1]-b[1])
        if 0 <= pos[0] < width and 0 <= pos[1] < height:
            antinodes1.add(pos)

antinodes2 = set()
for locations in antennas.values():
    for (a, b) in permutations(locations, 2):
        diff = (a[0]-b[0], a[1]-b[1]) # b to a
        pos = a
        while 0 <= pos[0] < width and 0 <= pos[1] < height:
            antinodes2.add(pos)
            pos = (pos[0]+diff[0], pos[1]+diff[1])

print(len(antinodes1))
print(len(antinodes2))

File: 2024/Day 08/C++/main.cpp
#include <string>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <utility>

int main(int argc, char *argv[])
{
    std::string line;
    int y = 0;
    std::unordered_map<char, std::vector<std::pair<int, int>>> antennas;
    while (std::getline(std::cin, line))
    {
        for (int x = 0; x < line.length(); x++)
        {
            if (line[x] != '.')
            {
                antennas[line[x]].push_back(std::make_pair(x, y));
            }
        }
        y++;
    }
    const int width = line.length();
    const int height = y;

    std::unordered_set<int> antinodes1;
    std::unordered_set<int> antinodes2;
    for (auto const &[key, locations] : antennas)
    {
        for (std::pair<int, int> a : locations)
        {
            for (std::pair<int, int> b : locations)
            {
                if (a != b)
                {
                    std::pair<int, int> pos1 = std::make_pair(2 * a.first - b.first, 2 * a.second - b.second);
                    if (0 <= pos1.first && pos1.first < width && 0 <= pos1.second && pos1.second < height)
                    {
                        antinodes1.insert(pos1.first + pos1.second * width);
                    }

                    std::pair<int, int> diff = std::make_pair(a.first - b.first, a.second - b.second);
                    std::pair<int, int> pos2 = a;
                    while (0 <= pos2.first && pos2.first < width && 0 <= pos2.second && pos2.second < height)
                    {
                        antinodes2.insert(pos2.first + pos2.second * width);
                        pos2 = std::make_pair(pos2.first + diff.first, pos2.second + diff.second);
                    }
                }
            }
        }
    }
    std::cout << antinodes1.size() << "\n";
    std::cout << antinodes2.size() << "\n";
}